cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -947,18 +947,25 @@ class AsofJoinNode : public ExecNode {
     if (state_.at(0)->Finished()) {
       ErrorIfNotOk(plan_->ScheduleTask([this] {
         outputs_[0]->InputFinished(this, batches_produced_);
+        finished_.MarkFinished();
```

Comment:
What if the previous line throws an exception? This line potentially calls into a node not provided by Arrow, so it could throw. The call to `MarkFinished` should happen in any case.

Reply:
I take a safety point of view here. Even if the method is not allowed to throw an exception, it is not something Arrow QA can validate, because the node called into is potentially outside of Arrow code. I'm proposing a small fix that prevents a developer oversight, due to causing an exception where one shouldn't occur, from escalating into a deadlock. A deadlock is strictly worse at least when a long-running execution is planned because the user may notice the deadlock much later than when it occurred; an early-failure is better.

Reply:
I agree an early-failure is better - an intended exception should not cause the system to deadlock ideally. @rtpsw I think I am missing sth - do you have a proposed fix?

Reply:
I'd propose fixing by wrapping `finished_.MarkFinished()` in a destructor for its scope - something like:
```
ErrorIfNotOk(plan_->ScheduleTask([this] {
        Defer cleanup([]() { finished_.MarkFinished(); });
        outputs_[0]->InputFinished(this, batches_produced_);
        return Status::OK();
      }));
```
where `Defer` is something like:
```
template <typename Callable>
struct Defer {
  Callable callable;
  Defer(Callable callable) : callable(std::move(callable)) {}
  ~Defer() noexcept { callable(); }
};
```

Reply:
Sorry for posting code here. I'll only free up to preparing a commit a bit later.